File: src/MazeGeneration.java
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Stack;

public class MazeGeneration {
	
	//enum direction for each cell to hold
	enum Direction{
		UP, DOWN, LEFT, RIGHT
	}
	
	public int width, height; //width height of maze
	private HashSet<Direction>[][] maze; //2d array of direction sets to represent maze 
	private int[] start; //start cell
	private HashSet<List<Integer>> visited, pEndpoints; //visited and potential end points list
	private Stack<int[]> backtrace; //backtrace stack
	private Stack<Direction> backtraceDir; //backtrace direction stack
	
	public void draw() {
//		System.out.println("DRAW: "); //prints values of every cell 
//		for (int i=0; i<height; i++) {
//			for (int j=0; j<width; j++) {
//				System.out.print("["+i+", "+j+"] ");
//				for (Direction dir: maze[i][j])
//					System.out.print(dir+" ");
//				System.out.println();
//			}
//		}
		for (int i=0; i<width; i++)  
			System.out.print(" __"); //prints the top of the maze
		System.out.println();
		for (int i=0; i<height; i++) {
			System.out.print('|'); //prints left wall of the maze
			for (int j=0; j<width; j++) {
				if(!maze[i][j].contains(Direction.DOWN))
					System.out.print("__"); 
				else
					System.out.print("  ");
				if(!maze[i][j].contains(Direction.RIGHT))
					System.out.print('|');
				else
					System.out.print(" ");
			} //draws maze
			System.out.println();
		}
	}

	public void generate() {
		generate(start, false);
		System.out.println(pEndpoints); //prints potential endpoints for the maze
		draw(); 
	}
	
	private void generate(int[] curr, boolean isDeadend) {
		if (Arrays.equals(curr, start)&&visited.size()!=0) { 
			return; //if it backtraces back to start, end generation
		}
		ArrayList<Direction> possibles=possibles(curr); //all possible directions from curr
		if (possibles.size()==0) { //if deadend or backtracing
			if (!isDeadend) pEndpoints.add(List.of(curr[0], curr[1])); //add to potential endpoints if true deadend
			visited.add(List.of(curr[0], curr[1])); //add point to visited
			maze[curr[0]][curr[1]].add(backtraceDir.pop()); //add previous direction
//			System.out.println("["+curr[0]+", "+curr[1]+"]"); //print backtrace steps
			generate(backtrace.pop(), true); //go backwards by popping from stack
			return;
		}
		Direction dir=possibles.get((int)(Math.random()*possibles.size())); //gets random valid direction
		visited.add(List.of(curr[0], curr[1])); //adds to visited set
		backtrace.push(curr); //pushes to backtrace
		maze[curr[0]][curr[1]].add(dir); //adds direction to cell
//		System.out.println("["+curr[0]+", "+curr[1]+"] "+dir); //prints step 
		switch(dir) { //pushes opposite direction for backtrace and moves to next cell
		case UP: 
			backtraceDir.push(Direction.DOWN); 
			generate(new int[] {curr[0]-1, curr[1]}, false);
			break;
		case DOWN:
			backtraceDir.push(Direction.UP);
			generate(new int[] {curr[0]+1, curr[1]}, false);
			break;
		case LEFT:
			backtraceDir.push(Direction.RIGHT);
			generate(new int[] {curr[0], curr[1]-1}, false);
			break;
		case RIGHT:
			backtraceDir.push(Direction.LEFT);
			generate(new int[] {curr[0], curr[1]+1}, false);
			break;
		}
	}
	
	private ArrayList<Direction> possibles(int[] n){ //creates list of possible directions 
		ArrayList<Direction> possibles=new ArrayList<Direction>();
		for (Direction dir: Direction.values())
			switch(dir) {
			case UP:
				if (n[0]==0||visited.contains(List.of(n[0]-1, n[1]))) break;
				possibles.add(dir);
				break;
			case DOWN:
				if (n[0]==height-1||visited.contains(List.of(n[0]+1, n[1]))) break;
				possibles.add(dir);
				break;
			case LEFT:
				if (n[1]==0||visited.contains(List.of(n[0], n[1]-1))) break;
				possibles.add(dir);
				break;
			case RIGHT:
				if (n[1]==width-1||visited.contains(List.of(n[0], n[1]+1))) break;
				possibles.add(dir);
				break;
			} 
		return possibles;
	}
	
	public MazeGeneration(int width, int height) { //constructor initialize all variables and calls generate
		this.width=width;
		this.height=height;
		maze=new HashSet[height][width];
		for (int i=0; i<height; i++) 
			for (int j=0; j<width; j++) 
				maze[i][j]=new HashSet<Direction>();
		start=new int[]{(int) (Math.random()*height), (int) (Math.random()*width)};
		visited=new HashSet<List<Integer>>();
		pEndpoints=new HashSet<List<Integer>>();
		backtrace=new Stack<int[]>();
		backtraceDir=new Stack<Direction>();
		generate(); //generate maze
		MazeSaver saver=new MazeSaver(); //save maze to text file
		try {
			saver.save("maze.txt", maze);
		} catch (IOException e) {
			e.printStackTrace();
		}
		MazeLoader loader=new MazeLoader(); //load maze from text file
		try {
			maze=loader.load("maze.txt");
			draw();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	
	public MazeGeneration(int n) { //constructor for square mazes
		this(n,n);
	}
	
	public static void main(String[] args) { //main method to create maze generator
		new MazeGeneration(45);
	}
}

File: src/MazeGeneration.cpp
#include "MazeGeneration.h"

MazeGeneration::MazeGeneration(int width, int height) {//constructor initialize all variables and calls generate
	srand(time(0));
	this->width = width;
	this->height = height;
	maze = new unordered_set<Direction>*[height];
	for (int i = 0; i < height; i++) {
		maze[i] = new unordered_set<Direction>[width];
	}
	start = {(int)(rand() % height), (int)(rand() % width)};
	MazeGeneration::visited = new unordered_set<vector<int>, VectorHash>(); 
	MazeGeneration::pEndpoints = new unordered_set<vector<int>, VectorHash>();
	backtrace = new stack<vector<int>>();
	backtraceDir = new stack<Direction>();
	generate(); //generate maze
}

void MazeGeneration::generate() {
	generate(start, false);
	draw();
	//cout < < pEndpoints; //prints potential endpoints for the maze
}

void MazeGeneration::draw() {
	for (int i=0; i<width; i++) {
		cout<<" __";
	}
	cout<<"\n";
	for (int i=0; i<height; i++) {
		cout<<"|";
		for (int j=0; j<width; j++){
			if(maze[i][j].find(DOWN)==maze[i][j].end())
				cout<<"__";
			else
				cout<<"  ";
			if (maze[i][j].find(RIGHT)==maze[i][j].end())
				cout<<"|";
			else 
				cout<<" ";
		}
		cout<<"\n";
	}
}


void MazeGeneration::generate(vector<int> curr, bool isDeadend) {
	if (curr == start && !visited->empty()) {
		return; //if it backtraces back to start, end generation
	}
	vector<Direction> possibles = *MazeGeneration::possibles(curr); //all possible directions from curr
	if (possibles.size() == 0) {//if deadend or backtracing
		if (!isDeadend) 
			MazeGeneration::pEndpoints->insert(curr); //add to potential endpoints if true deadend
		MazeGeneration::visited->insert(curr); //add point to visited
		Direction temp1 = backtraceDir->top();
		backtraceDir->pop();
		maze[curr.at(0)][curr.at(1)].insert(temp1);
		vector<int> temp2 = backtrace->top();
		backtrace->pop();
		MazeGeneration::generate(temp2, true); //go backwards by popping from stack
		return;
	}
	Direction dir = possibles.at(rand()%possibles.size()); //gets random valid direction
	visited->insert(curr); //adds to visited set
	backtrace->push(curr); //pushes to backtrace stack
	maze[curr.at(0)][curr.at(1)].insert(dir); //adds direction to cell
	switch (dir) {//pushes opposite direction for backtrace and moves to next cell
	case UP:
		backtraceDir->push(DOWN);
		MazeGeneration::generate({curr.at(0)-1, curr.at(1)}, false);
		break;
	case DOWN:
		backtraceDir->push(UP);
		MazeGeneration::generate({curr.at(0) + 1, curr.at(1)}, false);
		break;
	case LEFT:
		backtraceDir->push(RIGHT);
		MazeGeneration::generate({curr.at(0), curr.at(1) - 1}, false);
		break;
	case RIGHT:
		backtraceDir->push(LEFT);
		MazeGeneration::generate({curr.at(0), curr.at(1) + 1}, false);
		break;
	}
}

vector<MazeGeneration::Direction>* MazeGeneration::possibles(vector<int> n) {//creates list of possible directions 
	vector<MazeGeneration::Direction>* possibles = new vector<MazeGeneration::Direction>();
	if (!(n.at(0) == 0 || visited->find({n.at(0) - 1,n.at(1)}) != visited->end()))
		possibles->push_back(UP);
	if (!(n.at(0) == MazeGeneration::height - 1 || visited->find({n.at(0) + 1, n.at(1)}) != visited->end()))
		possibles->push_back(DOWN);
	if (!(n.at(1) == 0 || visited->find({n.at(0), n.at(1) - 1}) != visited->end()))
		possibles->push_back(LEFT);
	if (!(n.at(1) == width - 1 || visited->find({n.at(0), n.at(1) + 1}) != visited->end()))
		possibles->push_back(RIGHT);
	return possibles;
}

map<bool, vector<int>> MazeGeneration::DirectionToCoords(int depth) {
	vector<int> v, h;
	for (int i=0; i<height; i++) {
		for (int j=0; j<width; j++) {
			if (i==0&&!(maze[i][j].find(UP)!=maze[i][j].end())){
				h.push_back(j*depth);
				h.push_back(i*depth);
			}
			if (!(maze[i][j].find(DOWN)!=maze[i][j].end())){
				h.push_back(j*depth);
				h.push_back((i+1)*depth);
			}
			if (j==0&&!(maze[i][j].find(LEFT)!=maze[i][j].end())){
				v.push_back(j*depth); 
				v.push_back(i*depth);
			}
			if (!(maze[i][j].find(RIGHT)!=maze[i][j].end())){
				v.push_back((j+1)*depth);
				v.push_back(i*depth);
			}
		}
	}
	map<bool, vector<int>> map;
	map[false]=h;
	map[true]=v;
	cout<<"horizontal = { ";
	for (int z: map.at(false))
		cout<<z<<", ";
	cout<<"}\n";
	cout<<"vertical = { ";
	for (int z: map.at(true))
		cout<<z<<", ";
	cout<<"}\n";
	return map;
}

// MazeGeneration::MazeGeneration(int n) {//constructor for square mazes
// 	MazeGeneration(n, n);
// }

MazeGeneration::~MazeGeneration() {
	for (int i = 0; i < height; i++) {
		delete[] maze[i];
	}
	delete[] maze;
}

int main() {
	MazeGeneration obj(3, 3);
	obj.DirectionToCoords(10);
}